crypto/timestamp/timestamptest/tsa.go

```diff
@@ -0,0 +1,269 @@
+// Package timestamptest provides utilities for timestamp testing
+package timestamptest
+
+import (
+	"context"
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"math"
+	"math/big"
+	"time"
+
+	"github.com/notaryproject/notation-go-lib/crypto/timestamp"
+	"github.com/notaryproject/notation-go-lib/internal/crypto/cms"
+	"github.com/notaryproject/notation-go-lib/internal/crypto/hashutil"
+	"github.com/notaryproject/notation-go-lib/internal/crypto/oid"
+	"github.com/notaryproject/notation-go-lib/internal/crypto/pki"
+)
+
+// responseRejection is a general response for request rejection.
+var responseRejection = &timestamp.Response{
+	Status: pki.StatusInfo{
+		Status: pki.StatusRejection,
+	},
+}
+
+// TSA is a Timestamping Authority for testing purpose.
+type TSA struct {
+	// key is the TSA signing key.
+	key *rsa.PrivateKey
+
+	// cert is the self-signed certificate by the TSA signing key.
+	cert *x509.Certificate
+
+	// NowFunc provides the current time. time.Now() is used if nil.
+	NowFunc func() time.Time
+}
+
+// NewTSA creates a TSA with random credentials.
+func NewTSA() (*TSA, error) {
+	// generate key
+	key, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, err
+	}
+
+	// generate certificate
+	serialNumber, err := rand.Int(rand.Reader, big.NewInt(math.MaxInt64))
+	if err != nil {
+		return nil, err
+	}
+	now := time.Now()
+	template := x509.Certificate{
+		SerialNumber: serialNumber,
+		Subject: pkix.Name{
+			CommonName: "timestamp test",
+		},
+		NotBefore:             now,
+		NotAfter:              now.Add(365 * 24 * time.Hour), // 1 year
+		KeyUsage:              x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageTimeStamping},
+		BasicConstraintsValid: true,
+	}
+	certBytes, err := x509.CreateCertificate(rand.Reader, &template, &template, key.Public(), key)
+	if err != nil {
+		return nil, err
+	}
+	cert, err := x509.ParseCertificate(certBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return &TSA{
+		key:  key,
+		cert: cert,
+	}, nil
+}
+
+// Certificate returns the certificate used by the server.
+func (tsa *TSA) Certificate() *x509.Certificate {
+	return tsa.cert
+}
+
+// Timestamp stamps the time with the given request.
+func (tsa *TSA) Timestamp(_ context.Context, req *timestamp.Request) (*timestamp.Response, error) {
+	// validate request
+	if req.Version != 1 {
+		return responseRejection, nil
+	}
+	hash, ok := oid.ConvertToHash(req.MessageImprint.HashAlgorithm.Algorithm)
+	if !ok {
+		return responseRejection, nil
+	}
+	if hashedMessage := req.MessageImprint.HashedMessage; len(hashedMessage) != hash.Size() {
+		return responseRejection, nil
+	}
+
+	// generate token info
+	policy := asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 4146, 2} // time-stamp-policies
+	switch hash {
+	case crypto.SHA1:
```

Comment:
Does 2 or 3 mean anything? Or is it just a random number?

Reply:
Yes, they means diffrent policies.
See https://www.globalsign.com/en/repository/GlobalSign_CP_v6.7_final.pdf at page 18.